include/alpaka/math/Complex.hpp

```diff
@@ -0,0 +1,540 @@
+/* Copyright 2022 Sergei Bastrakov
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/Common.hpp>
+#include <alpaka/math/FloatEqualExact.hpp>
+
+#include <cmath>
+#include <complex>
+#include <iostream>
+
+namespace alpaka
+{
+    //! Implementation of a complex number useable on host and device.
+    //!
+    //! It follows the layout of std::complex and so array-oriented access.
+    //! Inside the class template implements all methods and operators as std::complex<T>.
+    //! Additionally, provides an implicit conversion to and from std::complex<T>.
+    //! All methods besides operators << and >> are host-device.
+    //! Note that it does not provide non-member functions of std::complex besides the operators.
+    //! Those are provided the same way as alpaka math functions for real numbers.
+    //!
+    //! Note that unlike most of alpaka, this is a concrete type template, and not merely a concept.
+    //!
+    //! Naming and order of the methods match https://en.cppreference.com/w/cpp/numeric/complex.
+    //!
+    //! @tparam T type of the real and imaginary part: float, double, or long double.
+    template<typename T>
+    class Complex
+    {
+    public:
+        //! Type of the real and imaginary parts
+        using value_type = T;
+
+        //! Constructor from the given real and imaginary parts
+        constexpr ALPAKA_FN_HOST_ACC Complex(T const& real = T{}, T const& imag = T{}) : m_real(real), m_imag(imag)
+        {
+        }
+
+        //! Copy constructor
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex const& other) = default;
+
+        //! Constructor from Complex of another type
+        template<typename U>
+        constexpr ALPAKA_FN_HOST_ACC Complex(Complex<U> const& other)
+            : m_real(static_cast<T>(other.real()))
+            , m_imag(static_cast<T>(other.imag()))
+        {
+        }
```

Comment:
```suggestion
        template<typename U>
        explicit constexpr ALPAKA_FN_HOST_ACC Complex(Complex<U> const& other)
            : m_real(static_cast<T>(other.real()))
            , m_imag(static_cast<T>(other.imag()))
        {
        }
```
You may consider making conversion between complex numbers of different FP type explicit.

Reply:
It's non-explicit [here](https://en.cppreference.com/w/cpp/numeric/complex/complex), so i did the same

Reply:
Fine!